[Mempool] Improve removal of invalid transactions after reorgs

sdaftuar opened the pull request:
@TheBlueMatt This is a rebased version of #6656 (which in turn built off #6595).

As previously reported in #6595, there's currently a bug where we don't remove transactions that become invalid after a reorg due to the transaction's locktime.  The solution proposed in #6595 doesn't exhibit great behavior because it doesn't make sense to evict locktime-invalid transactions in the middle of a reorg before the new tip is updated (since typically the block height is the same or greater at the end); @TheBlueMatt addressed this in #6656, which is rebased and included here.

I have also added a commit that vastly improves the efficiency of `CTxMemPool::removeForReorg()`, by tracking in each mempool entry whether or not the transaction spends a coinbase. Then in `removeForReorg()`, we only access coins for those transactions that spend a coinbase, instead of looking up all inputs for all transactions in the mempool (which is slow and blows up the UTXO cache).

I compared this code with and without this last commit on some historical data, and observed that this drastically improves runtime of `removeForReorg()` and reduces memory usage in `pcoinsTip`.  Running this code on historical data on a few days in July, I observed 4 reorgs, and the runtime of `removeForReorg()` went from ranging between 167ms - 719ms down to between 5.5ms-16ms; meanwhile `pcoinsTip` memory usage (after the call to `removeForReorg()`) was reduced by between 144MB - 413MB.

When BIP68 support is merged we'll probably want to update this approach, since `IsFinalTx()` will become a more expensive function that needs to look at inputs.

TheBlueMatt committed:
Add failing test checking timelocked-txn removal during reorg

TheBlueMatt committed:
Fix removal of time-locked transactions during reorg

TheBlueMatt committed:
Fix comment in removeForReorg

TheBlueMatt committed:
Make indentation in ActivateBestChainStep readable

TheBlueMatt committed:
removeForReorg calls once-per-disconnect-> once-per-reorg

sdaftuar committed:
Track coinbase spends in CTxMemPoolEntry

This allows us to optimize CTxMemPool::removeForReorg.

dcousens commented:
concept ACK

TheBlueMatt commented:
Didnt look too closely, but the last commit seems reasonable.

petertodd commented:
Needs rebase

sdaftuar commented:
Rebased

sipa commented:
utACK.

For BIP68 transactions maybe it makes sense to replace the spendscoinbase flag with a minimum height/time at which the transaction becomes valid (which also generalized the wanted behaviour for coinbase-spending transactions).

sdaftuar committed:
Don't call removeForReorg if DisconnectTip fails

morcos commented:
@sipa, @sdaftuar and I discussed adding the BIP68 time and locktime at length.
We were waiting for the code to finalize, and had gotten hung up on the locktime potentially being quite recent and therefore causing a great many transactions to have all 3 heights recalculated on a reorg.  But now I'm realizing it makes a lot of sense to only include BIP68 and coinbase height, and then separately scan for locktime since its so easy!

In any case, I think this should be merged and that should be a future improvement.
I will review shortly.

sdaftuar commented:
Nits fixed.

morcos commented:
utACK

sdaftuar commented:
Rebased

dcousens commented:
Easiest to read with [`?w=0`](https://github.com/bitcoin/bitcoin/pull/6915/files?w=0) (omits white space changes).

dcousens commented:
once-over utACK

gmaxwell commented:
I think this needs an update for MTP, as it appears to be using adjusted time. @TheBlueMatt

sdaftuar commented:
@gmaxwell Oops -- thanks for catching that.  I thought it might make sense to pull `removeForReorg` out of the mempool and into main, so that we can keep policy code more organized and hopefully duplicate less logic.

If this approach seems okay, I'll try to go back and clean up the commit history here.

ping @TheBlueMatt

TheBlueMatt commented:
I'm not a huge fan. This isnt actually policy code, its code who's behavior is derived from consensus rules, so I dont think it belongs in main. I'd much rather pass the MTP spendable-height to CTxMemPool::...() than to do this all in main.

sdaftuar committed:
Fix removeForReorg to use MedianTimePast

sdaftuar commented:
Ok I tried the smaller change of just passing through the MTP into removeForReorg, is this what you had in mind?

sdaftuar commented:
>18:22 < BlueMatt> anyway, I'd prefer to not duplicate the code....can we pull that out 
>                 into a function?

@TheBlueMatt: Yeah I also didn't want to duplicate that code, but it seemed like the right way to do that would be to reuse the `CheckFinalTx` function, which does very little, here's what it looks like:

```
bool CheckFinalTx(const CTransaction &tx, int flags)
{
    AssertLockHeld(cs_main);

    // By convention a negative value for flags indicates that the
    // current network-enforced consensus rules should be used. In
    // a future soft-fork scenario that would mean checking which
    // rules would be enforced for the next block and setting the
    // appropriate flags. At the present time no soft-forks are
    // scheduled, so no flags are set.
    flags = std::max(flags, 0);

    // CheckFinalTx() uses chainActive.Height()+1 to evaluate
    // nLockTime because when IsFinalTx() is called within
    // CBlock::AcceptBlock(), the height of the block *being*
    // evaluated is what is used. Thus if we want to know if a
    // transaction can be part of the *next* block, we need to call
    // IsFinalTx() with one more than chainActive.Height().
    const int nBlockHeight = chainActive.Height() + 1;

    // Timestamps on the other hand don't get any special treatment,
    // because we can't know what timestamp the next block will have,
    // and there aren't timestamp applications where it matters.
    // However this changes once median past time-locks are enforced:
    const int64_t nBlockTime = (flags & LOCKTIME_MEDIAN_TIME_PAST)
                             ? chainActive.Tip()->GetMedianTimePast()
                             : GetAdjustedTime();

    return IsFinalTx(tx, nBlockHeight, nBlockTime);
}
```

We're basically duplicating all the code here (except maybe that `std::max(flags, 0)`; I don't really understand what that is) already, so pulling out the one line to avoid duplicating that doesn't seem worth it...  I'd rather find a way to reuse the whole function (which is what I did in my first attempt).

My only other idea here would be to change `CTxMemPool::removeForReorg` to take a "flags" argument only, and then invoke `CheckFinalTx` instead of `IsFinalTx`.

NicolasDorier commented:
in #6312 IsFinalTx (renamed LockTime) is responsible for calculating the blocktime.
LockTime function can be called by removeForReorg with the passed flag, this would prevent code duplication.

sipa commented:
Needs rebase.

sdaftuar commented:
Rebased, and I changed the last commit to call `CheckFinalTx`.  I don't love this, but it avoids the code duplication of having all the callers separately calculate what time to pass in.

@NicolasDorier  We'll want to reimplement this function after #6312, since `LockTime` will be a more expensive function to call.  I'm happy to tackle that when the time comes, but I don't want to rebase this on #6312 since this is tagged for 0.12 and fixes a bug.  For now, I think this formulation should be easy to merge into #6312, since I'm now just passing the flags variable through to `CheckFinalTx`.
